lib/components/admin-command.js

```diff
@@ -0,0 +1,157 @@
+        });
+
+        if (missing.length) {
+            throw Error("Missing required options: " + missing.join(", "));
+        }
+
+        args = opts._;
+        delete opts["_"];
+    }
+
+    return this._func(bridge, opts, args, respond);
+};
+
+/**
+ * Helper method to construct a 'help' AdminCommand instance
+ * @param {Object} commands The object containing the application's
+ * AdminCommand instances.
+ * @return {AdminCommand} A new AdminCommand.
+ *
+ * Typically this method would be invoked just after the admin command object
+ * is declared, to provide the automatic online help command.
+ *
+ *   var commands = {};
+ *   commands.help = AdminCommand.makeHelpCommand(commands);
+ */
+AdminCommand.makeHelpCommand = function(commands) {
+    return new AdminCommand({
+        desc: "display a list of commands",
+        func: function(bridge, _, args, respond) {
+            if (args.length == 0) {
+                Object.keys(commands).sort().forEach(function (k) {
+                    var cmd = commands[k];
+                    respond(k + ": " + cmd.desc);
+                });
+            }
+            else {
+                var name = args.shift();
+                var cmd = commands[name];
+                if (!cmd) {
+                    throw Error("No such command '" + name + "'");
+                }
+
+                respond(name + " - " + cmd.desc);
+                var optspec = cmd.optspec || {};
+                Object.keys(optspec).sort().forEach((n) => {
+                    respond("  --" + n + ": " + optspec[n].desc);
+                });
+            }
+        },
+    });
+};
+
+/**
+ * A specification of command-line options recognised by an AdminCommand.
+ * @typedef {Object} AdminCommand~opts
+ * Each option is given as a key to this object whose key name is the name
+ * of the option, and the value is a string giving some descriptive text for
+ * the 'help' command.
+ * Each option name may be suffixed by aliases (for example, single letter
+ * short forms) separated by pipe symbols ('|'). Users may enter options in
+ * any of these shortcuts; the value presented to the command handler function
+ * will come in a key named after the first name component.
+ * Each option may be declared as mandatory, by prefixing the name string with
+ * an exclamation point ('!'). It then becomes an error to try to invoke the
+ * command without providing a value for that option; effectively turning it
+ * into a named argument instead.
+ */
+
+/**
+ * The handler function for an AdminCommand that contains its actual
+ * implementation code.
+ * @typedef {function} AdminCommand~func
+ * @param {Object} bridge The top-level bridge instance.
+ * @param {Object} opts Values of any commandline options the user supplied
+ * @param {string[]} args Remaining positional commandline arguments after the
+ * options have been parsed out.
+ * @param {function} respond The responder function that the command handler
```

Comment:
In no place do we specify what this looks like - how is it invoked? Does it have any return args?